File: src/main/java/nieboczek/lifestolen/serializer/base/DoubleSerializer.java
package nieboczek.lifestolen.serializer.base;

import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder;
import nieboczek.lifestolen.serializer.lang.TokenStream;
import nieboczek.lifestolen.serializer.lang.TokenType;

public class DoubleSerializer extends Serializer<Double> {
    private DoubleSerializer() {
    }

    public static DoubleSerializer of() {
        return new DoubleSerializer();
    }

    @Override
    public void serialize(Double value, SerializedStringBuilder builder) {
        builder.text(value.toString());
    }

    @Override
    public Double deserialize(TokenStream stream) {
        return Double.parseDouble(stream.nextTokenText(TokenType.NUMBER));
    }
}

File: src/main/kotlin/nieboczek/lifestolen/serializer/base/DoubleSerializer.kt
package nieboczek.lifestolen.serializer.base

import nieboczek.lifestolen.serializer.SerializerError
import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder
import nieboczek.lifestolen.serializer.lang.TokenStream
import nieboczek.lifestolen.serializer.lang.TokenType

class DoubleSerializer : Serializer<Double>() {
    override fun serialize(value: Double, builder: SerializedStringBuilder) {
        builder.text(value.toString())
    }

    override fun deserialize(stream: TokenStream): Double {
        val text = stream.nextTokenText(TokenType.NUMBER)
        try {
            return text.toDouble()
        } catch (e: NumberFormatException) {
            throw SerializerError("Invalid double value: \"$text\"", e)
        }
    }
}
